### ISSUE TITLE
cv2 module members are not recognized

### ISSUE THREAD

**Issue by uriva** [PyCQA/pylint#2426]
When importing cv2 I get errors such as:

[pylint] E1101:Module 'cv2' has no 'imread' member
[pylint] E1101:Module 'cv2' has no 'resize' member
[pylint] E1101:Module 'cv2' has no 'imshow' member
[pylint] E1101:Module 'cv2' has no 'waitKey' member
[pylint] E1101:Module 'cv2' has no 'destroyAllWindows' member

etc'

(the code itself runs fine)

I'm using VSCode 1.26.1 and Python 3.6.5.

See this issue for more complaints: https://github.com/DonJayamanne/pythonVSCode/issues/623

**Comment by PCManticore** [comment]
This should work if you pass `--extension-pkg-whitelist=cv2` to pylint as in `pylint --extension-pkg-whitelist=cv2 ...`. The reason for this is that cv2 seems to be an extension package, thus it's C code that pylint cannot understand. The extension whitelist flag lets pylint import `cv2` in order to build an AST out of the imported cv2 module.

**Comment by uriva** [comment]
To future readers, in vsc this means adding this line to the settings:
"python.linting.pylintArgs": ["--extension-pkg-whitelist=cv2"]

**Comment by cipri-tom** [comment]
Thanks @PCManticore 

However, even with the extension enables, pylint still doesn't recognise 2nd level members, such as `cv2.ml.KNearest_create()` or `cv2.ximgproc.thinning()`. 

I believe it is the same case as #779 ?

**Comment by ashraf-minhaj** [comment]
> python.linting.pylintArgs": ["--extension-pkg-whitelist=cv2"]

How do I add this in settings? I can't find anything. Actually I'm a beginner at all these

**Comment by cipri-tom** [comment]
@ashraf-minhaj use the Cmd+shift+p shortcut (or ctrl instead of cmd) and type "open settings (json)". There should be an option which opens the json settings and you can add the line there

**Comment by Davidnet** [comment]
I'm encountering the same situation: 

```
dgromov@Predator-G3-571 ~/D/u/s/p/CarND-Advanced-Lane-Lines> pylint camera_calibration.py --extension-pkg-whitelist=cv2 | tail                                                                                                       (SDCND) 
camera_calibration.py:157:18: E1101: Module 'cv2' has no 'undistort' member (no-member)
camera_calibration.py:159:12: E1101: Module 'cv2' has no 'imshow' member (no-member)
camera_calibration.py:160:12: E1101: Module 'cv2' has no 'waitKey' member (no-member)
camera_calibration.py:162:12: E1101: Module 'cv2' has no 'imwrite' member (no-member)
camera_calibration.py:163:12: E1101: Module 'cv2' has no 'destroyAllWindows' member (no-member)
camera_calibration.py:129:0: R1711: Useless return at end of function or method (useless-return)

----------------------------------------------------------------------
Your code has been rated at -12.67/10 (previous run: -12.67/10, +0.00)

dgromov@Predator-G3-571 ~/D/u/s/p/CarND-Advanced-Lane-Lines> conda list | grep  "pylint\|astroid"                                                                                                                                    (SDCND) 
astroid                   2.1.0                    py36_0  
pylint                    2.2.2                    py36_0  
```

**Comment by PCManticore** [comment]
As mentioned, most of the errors should be already solved by using ``--extension-pkg-whitelist=cv2``. But unfortunately, as @cipri-tom mentioned, there are a couple of ``cv2`` modules for which that flag does not work. I think the best way forward is for someone to propose a patch to `astroid`, the library that `pylint` uses for infererence, to add support for `cv2` in a similar fashion to how we added for `numpy` in #779. Unfortunately we're not familiar with `cv2` so it's not something that we'd do ourselves.

**Comment by mehanton** [comment]
https://www.bilibili.com/video/av33693850/
"python.linting.pylintArgs": ["--generate-members"]

**Comment by fuzzyBatman** [comment]
For me, pylint cannot process cv2 even with `"python.linting.pylintArgs": ["--generate-members"]` when I build cv2 from source. Works when installed using pip.

**Comment by yeknafar** [comment]
@mehanton this works for me. 

Thank you. :)

**Comment by LintangWisesa** [comment]
1. On VScode: CTRL + Shift + P
2. Choose "Preferences: Open Settings (JSON)"
3. Add this line into JSON file: 
`"python.linting.pylintArgs": ["--generate-members"]`
4. Done, it works for me 👍

**Comment by githarshp** [comment]
> 1. On VScode: CTRL + Shift + P
> 2. Choose "Preferences: Open Settings (JSON)"
> 3. Add this line into JSON file:
>    `"python.linting.pylintArgs": ["--generate-members"]`
> 4. Done, it works for me 👍

Thanks, man! it worked for me too

**Comment by nyck33** [comment]
on VSCode Ubuntu I did "python.linting.pylintArgs": ["--generate-members", "--extension-pkg-whitelist=cv2"] which worked.

**Comment by insung3511** [comment]
> 1. On VScode: CTRL + Shift + P
> 2. Choose "Preferences: Open Settings (JSON)"
> 3. Add this line into JSON file:
>    `"python.linting.pylintArgs": ["--generate-members"]`
> 4. Done, it works for me 👍

Hm.... I can't edit "defaultSetting.json file....

**Comment by LintangWisesa** [comment]
> > 1. On VScode: CTRL + Shift + P
> > 2. Choose "Preferences: Open Settings (JSON)"
> > 3. Add this line into JSON file:
> >    `"python.linting.pylintArgs": ["--generate-members"]`
> > 4. Done, it works for me 👍
> 
> Hm.... I can't edit "defaultSetting.json file....

@insung3511 It must be __"settings.json"__, not __"defaultSetting.json"__! Make sure you choose __"Preferences: Open Settings (JSON)"__, not __"Preferences: Open Default Settings (JSON)"__ 👌

**Comment by duttanaman1** [comment]
thanks

**Comment by funny860** [comment]
thanks for the help guys

**Comment by Blinue** [comment]
The "--generate-members" option is wrong!
It works because pylint fails.
The command "pylint --generate-members test.py" fails because of "no such option: --generate-members".

**Comment by Blinue** [comment]
Look at [this](http://pylint.pycqa.org/en/latest/technical_reference/c_extensions.html), but the solution mentioned here won't work either.
It seems that the only solution is to disable the "no-member" error.

**Comment by rivergold** [comment]
The `"python.linting.pylintArgs": ["--generate-members"]` is wrong, it will make pylint not work.
The correct way is to set `""python.linting.pylintArgs": ["--generated-members=cv2.*"]` which only config for `cv2`, reference @bjtho08 reply in #779.

**Comment by t31060** [comment]
i compiled opencv from soruce and passed the args to pylitn but still intellisense doesnt work, there is another way to solve?

**Comment by guille-ds** [comment]
Check that "d" at the end of "--generate"...
This worked for me:

`"python.linting.pylintArgs": ["--generate-members=cv2.*"]`

**Comment by Saksham2dubey** [comment]
I also facing the same problem of "cv2 module members are not recognized
"
what I did was 
after importing cv2
write:
from cv2 import cv2
this worked for me because my JSON is not amendable

**Comment by JulianToledano** [comment]
> > 1. On VScode: CTRL + Shift + P
> > 2. Choose "Preferences: Open Settings (JSON)"
> > 3. Add this line into JSON file:
> >    `"python.linting.pylintArgs": ["--generate-members"]`
> > 4. Done, it works for me +1
> 
> Hm.... I can't edit "defaultSetting.json file....

Now my pylint just don't show any erros :sweat_smile:

**Comment by JulianToledano** [comment]
> The `"python.linting.pylintArgs": ["--generate-members"]` is wrong, it will make pylint not work.
> The correct way is to set `""python.linting.pylintArgs": ["--generated-members=cv2.*"]` which only config for `cv2`, reference @bjtho08 reply in #779.

Thanks that solves it.
 * On VScode: CTRL + Shift + P
 * Choose "Preferences: Open Settings (JSON)"
 * Add this line into JSON file: `"python.linting.pylintArgs": ["--generated-members=cv2.*"]`
 * Done, it works for me +1

**Comment by namhar** [comment]
I tried above and still no luck.

**Comment by akash-falcon1999** [comment]
> https://www.bilibili.com/video/av33693850/
> "python.linting.pylintArgs": ["--generate-members"]

works now thanks

**Comment by bjtho08** [comment]
> > The `"python.linting.pylintArgs": ["--generate-members"]` is wrong, it will make pylint not work.
> > The correct way is to set `""python.linting.pylintArgs": ["--generated-members=cv2.*"]` which only config for `cv2`, reference @bjtho08 reply in #779.
> 
> Thanks that solves it.
> 
> * On VScode: CTRL + Shift + P
> * Choose "Preferences: Open Settings (JSON)"
> * Add this line into JSON file: `"python.linting.pylintArgs": ["--generated-members=cv2.*"]`
> * Done, it works for me +1

@akash-falcon1999 that pylintarg is wrong, as you can see in the above. It will make any error and any warning from pylint disappear, even those that are supposed to be there. Trust me, you don't want to pass that pylintarg. Better to live with this bug, because it actually is a bug in pylint.

**Comment by akash-falcon1999** [comment]
> > > The `"python.linting.pylintArgs": ["--generate-members"]` is wrong, it will make pylint not work.
> > > The correct way is to set `""python.linting.pylintArgs": ["--generated-members=cv2.*"]` which only config for `cv2`, reference @bjtho08 reply in #779.
> > 
> > 
> > Thanks that solves it.
> > 
> > * On VScode: CTRL + Shift + P
> > * Choose "Preferences: Open Settings (JSON)"
> > * Add this line into JSON file: `"python.linting.pylintArgs": ["--generated-members=cv2.*"]`
> > * Done, it works for me +1
> 
> @akash-falcon1999 that pylintarg is wrong, as you can see in the above. It will make any error and any warning from pylint disappear, even those that are supposed to be there. Trust me, you don't want to pass that pylintarg. Better to live with this bug, because it actually is a bug in pylint.

@bjtho08  Thanks ,rectified it,was quite bugged off with those bugs,although the code was working fine.

**Comment by rishabh547** [comment]
> 1. On VScode: CTRL + Shift + P
> 2. Choose "Preferences: Open Settings (JSON)"
> 3. Add this line into JSON file:
>    `"python.linting.pylintArgs": ["--generate-members"]`
> 4. Done, it works for me 👍

Do we need to install json extension in vscode for this?

**Comment by bjtho08** [comment]
@rishabh547 see my comment above

**Comment by rishabh547** [comment]
> @rishabh547 see my comment above

So basically after adding that line in the file save the file right?  That's it?

**Comment by rishabh547** [comment]
Please help me I am low-key getting depressed because of this error, I have spent hours on this still unable to rectify!

**Comment by rishabh547** [comment]
> I tried above and still no luck.

Hey did you solve your issue? If so how

**Comment by bjtho08** [comment]
@rishabh547 you misunderstood. The line you wrote is wrong and will make pylint fail silently, thus making it appear as if your code has zero errors. You need to use the pylint key-value argument that is written in my comment. That will solve your issue without causing pylint to fail.

**Comment by ShivamMukherjee** [comment]
This doesn't seem to be bringing up `cv2` in the autocomplete suggestions.
Moreoever, hovering over cv2 shows this:
![image](https://user-images.githubusercontent.com/12777564/88017269-f52f0180-cb42-11ea-8075-31813e4a0502.png)
My built opencv from source, so this points to the `python-loader` module under "../opencv/build".
While the errors are all gone, no autocomplete suggestions come up.

**Comment by bjtho08** [comment]
@ShivamMukherjee autocomplete is not handled by pylint, but rather the languageserver. You will have to look at Microsofts languageserver (or jedi, if you use that one) for a solution.

**Comment by ShivamMukherjee** [comment]
@bjtho08 no dice there, sigh. It's funny that the REPL somehow does this easily.

**Comment by Shinit5201** [comment]
> To future readers, in vsc this means adding this line to the settings:
> `"python.linting.pylintArgs": ["--extension-pkg-whitelist=cv2"]`

how to add this line to settings???

**Comment by ShivamMukherjee** [comment]
@Shinit5201 
![image](https://user-images.githubusercontent.com/12777564/89611882-ea42d380-d89b-11ea-959c-92d43a821240.png)

**Comment by Shinit5201** [comment]
Thanks

**Comment by kshitij-agarwal** [comment]
After reading such complex tasks, I was totally surprised. 

A very simple way is: 
Instead of **`import cv2`** import **`from cv2 import cv2`**

It works like a charm.

**Comment by Vivraan** [comment]
> After reading such complex tasks, I was totally surprised.
> 
> A very simple way is:
> Instead of **`import cv2`** import **`from cv2 import cv2`**
> 
> It works like a charm.

You could also take a step further and import even more granularly into individual classes or features instead of the whole module!

**Comment by kritikaparmar-programmer** [comment]
> After reading such complex tasks, I was totally surprised.
> 
> A very simple way is:
> Instead of **`import cv2`** import **`from cv2 import cv2`**
> 
> It works like a charm.

It worked dude!!😁

**Comment by slugb0t** [comment]
> After reading such complex tasks, I was totally surprised.
> 
> A very simple way is:
> Instead of **`import cv2`** import **`from cv2 import cv2`**
> 
> It works like a charm.

a simple solution wow thank you

**Comment by zchrissirhcz** [comment]
`"python.linting.pylintArgs": ["--generate-members"]` works for me.

Btw, tts 2020-09-28 already, why people still have to setup this stuff for pylint?

**Comment by bjtho08** [comment]
> `"python.linting.pylintArgs": ["--generate-members"]` works for me.
> 
> Btw, tts 2020-09-28 already, why people still have to setup this stuff for pylint?

That argument is wrong and will make pylint fail silently. See the comments earlier in the thread for an explanation.

**Comment by jofrev** [comment]
> 
> 
> After reading such complex tasks, I was totally surprised.
> 
> A very simple way is:
> Instead of **`import cv2`** import **`from cv2 import cv2`**
> 
> It works like a charm.

It won't report an error in VS Code's problems tab any longer that's true. But if you run pylint from your terminal you'll still get

```
I1101: Module 'cv2.cv2' has no 'imread' member, but source is unavailable. Consider adding this module to extension-pkg-whitelist if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
```  

with **`from cv2 import cv2`** instead of

```bash
E1101: Module 'cv2' has no 'imread' member (no-member)
```

with **`import cv2`**. Depending on how one intends to use pylint, whitelisting **`cv2`** in pylint settings as described numerous times above may actually be better.

**Comment by gautamjain1009** [comment]
> @Shinit5201
> ![image](https://user-images.githubusercontent.com/12777564/89611882-ea42d380-d89b-11ea-959c-92d43a821240.png)

this is not working for me I am using Opencv 4.1.0 build from source in an anaconda environment.

**Comment by Mechazo11** [comment]
> 1. On VScode: CTRL + Shift + P
> 2. Choose "Preferences: Open Settings (JSON)"
> 3. Add this line into JSON file:
>    `"python.linting.pylintArgs": ["--generate-members"]`
> 4. Done, it works for me 👍

Worked for me too. Thanks

**Comment by ardf** [comment]
> I also facing the same problem of "cv2 module members are not recognized
> "
> what I did was
> after importing cv2
> write:
> from cv2 import cv2
> this worked for me because my JSON is not amendable

Thanks! worked for me too

**Comment by nnmrts** [comment]
I can't believe how many people here are recommending to edit the clumsy vscode setting `"python.linting.pylintArgs"`.

Not only is it just solving the issue locally, especially when you edit **your** global vscode `settings.json`, it also overrides any settings from a `.pylintrc` file, which is not immediately obvious. Using vscode's settings to add command line arguments to some binary running in the background is a bad idea in general, and should only be done as a last resort.

What you should **really** do, in my opinion, is to generate a pylint configuration file inside your project folder like this:

```zsh
pylint --generate-rcfile > .pylintrc
```

and then edit the settings `extension-pkg-allow-list` and `generated-members` there.

You can of course also edit these settings in your user- or system-wide .pylintrc, I'd advise against that though.

**Comment by domajstorovic** [comment]
> I can't believe how many people here are recommending to edit the clumsy vscode setting `"python.linting.pylintArgs"`.
> 
> Not only is it just solving the issue locally, especially when you edit **your** global vscode `settings.json`, it also overrides any settings from a `.pylintrc` file, which is not immediately obvious. Using vscode's settings to add command line arguments to some binary running in the background is a bad idea in general, and should only be done as a last resort.
> 
> What you should **really** do, in my opinion, is to generate a pylint configuration file inside your project folder like this:
> 
> ```shell
> pylint --generate-rcfile > .pylintrc
> ```
> 
> and then edit the settings `extension-pkg-allow-list` and `generated-members` there.
> 
> You can of course also edit these settings in your user- or system-wide .pylintrc, I'd advise against that though.

This. By far the most elegant solution. Thanks!

**Comment by KukavicaHome** [comment]
VS2019 works for me.

**Comment by ngvq** [comment]
install opencv_headless and it works for me.

**Comment by ZhiliangWu** [comment]
For `pyproject.toml` file, add the following 

```toml
[tool.pylint.typecheck]
generated-members = "cv2.*"
```

**Comment by Pierre-Sassoulas** [comment]
Note that using ``genrated-members`` should be the last solution if nothing else work, because it will disable pylint's check on everything coming from cv2.

**Comment by knowledgeLover** [comment]
pip installation is wrong. 

Try installing opencv-python-headless python dependency instead of opencv-python. That includes a precompiled binary wheel with no external dependencies (other than numpy), and is intended for headless environments like Docker. This saved almost 700mb in my docker image compared with using the python3-opencv Debian package (with all its dependencies).

The [package documentation](https://pypi.org/project/opencv-python/) discusses this and the related (more expansive) opencv-contrib-python-headless pypi package.

**Comment by DungGramer** [comment]
Update 2023:

```json
  "pylint.args": ["--extension-pkg-whitelist=cv2"]
```

**Comment by LambdaScorpii** [comment]
Hello There,

since the new release of OpenCV this no longer works. It still throws the no member error, regardless if you put in "--extension-pkg-whitelist=cv2" or --extension-pkg-allow-list=cv2 (as suggested in the official documentation, as "whitelist"seems dprecated). Any suggestion how to make that work? 

Generating all members defies the purpose of PyLint in my opinion...

**Comment by Pierre-Sassoulas** [comment]
Could you open a new issue with a reproducer please @knowledgeLover ?

**Comment by SajjadAemmi** [comment]
Add `--generated-members` argument when you run **pylint** command:

```bash
pylint --generated-members=cv2.* client.py
```

**Comment by LambdaScorpii** [comment]
hi @SajjadAemmi,

Thanks for the advice. Unfortubnately, this won't work well because, a) you annot configure this in a persistent manner in the pylint extenstion and b) this removes every scope on linting when it starts with cv2 regardless if right or wrong, so kind of defies the point of the whole linting.
